### PR TITLE
React Unit Test

### PR DESCRIPTION
**Pull request by channamakover**

### CODE REVIEW

File: 7-React/src/components/TodoItem/TodoItem.tsx [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+          <input data-hook={dataHooks.checkbox} type="checkbox" onClick={toggleHandler} className={isDone? classes['done-task']:'' }/>
+      <h3 data-hook={dataHooks.title}>{title}</h3>
+      <div className={classes["buttons-wrapper"]}>
+        <button
+          data-hook={dataHooks.editBtn}
+          className={classes["button-style"]}
+          onClick={editTaskHandler}
+        />
+        edit
+        <button />
```
Comment (tombenezra) [inline 7-React/src/components/TodoItem/TodoItem.tsx]: 🦄

File: 7-React/src/components/TodoItem/TodoItem.tsx [hunk 2]
```diff
@@ -0,0 +1,44 @@
+import React,{useState} from "react";
+import classes from "./TodoItem.module.css";
+import dataHooks from "./data-hooks";
+
+const TodoItem = ({
```
Comment (tombenezra) [inline 7-React/src/components/TodoItem/TodoItem.tsx]: describe the props of *every* component with explicit interface
```
const TodoItem: React.FC<Props> {...}
```

File: 7-React/src/components/TodoItem/TodoItem.tsx [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  title,
+  deleteTaskHandler,
+    editTaskHandler,
+}: {
+  title: string;
+  deleteTaskHandler: () => {};
```
Comment (tombenezra) [inline 7-React/src/components/TodoItem/TodoItem.tsx]: this should be mandatory, no? then there shouldn't be default values

File: 7-React/src/components/TodoItem/TodoItem.tsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  
+    }) => {
+    
+    const [isDone, setIsDone] = useState(false);
+    const toggleHandler = () => {
+        setIsDone(prevState => !prevState);
+        
+    }
+  return (
+    <li data-hook={dataHooks.root} className={classes["to-do-tasks"]}>
```
Comment (tombenezra) [inline 7-React/src/components/TodoItem/TodoItem.tsx:15]: even tho the class name is `to-do-tasks` you can access it with `classes.toDoTasks`. this will ensure that if someone might change it in the future, you will get error

File: 7-React/src/components/TodoItem/TodoItem.tsx [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+          <input data-hook={dataHooks.checkbox} type="checkbox" onClick={toggleHandler} className={isDone? classes['done-task']:'' }/>
```
Comment (tombenezra) [inline 7-React/src/components/TodoItem/TodoItem.tsx]: consider using [classnames](https://www.npmjs.com/package/classnames)

File: 7-React/src/components/TodoItem/TodoItem.spec.tsx [hunk 6]
```diff
@@ -0,0 +1,48 @@
+import TodoItemDriver from "./TodoItem.driver";
+import { Chance } from "chance";
+
+describe("Todo Item ", () => {
+  let driver: TodoItemDriver;
+  const chance = Chance();
```
Comment (tombenezra) [inline 7-React/src/components/TodoItem/TodoItem.spec.tsx:6]: you can initialize it outside the describe, just in case other describes want to use it

File: 7-React/src/components/TodoItem/TodoItem.driver.tsx [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+class TodoItemDriver {
+  private wrapper?: RenderResult;
+  private todoDescription: string = "";
+  private deleteTaskMock: jest.Mock = jest.fn();
+  private editTextMock: jest.Mock = jest.fn();
+  private toggleDone: jest.Mock = jest.fn();
+
+  private editBtn = () => {
+    const element = this.wrapper?.queryByTestId(dataHooks.editBtn);
```
Comment (tombenezra) [inline 7-React/src/components/TodoItem/TodoItem.driver.tsx]: check `this.wrapper` was initialized and throw error with proper message otherwise?

File: 7-React/src/components/TodoItem/TodoItem.driver.tsx [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    return {
+      editBtnExists: () => {
+        return element ? true : false;
```
Comment (tombenezra) [inline 7-React/src/components/TodoItem/TodoItem.driver.tsx]: specifically you can check if it's `null` or not

File: 7-React/src/components/TodoItem/TodoItem.driver.tsx [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+      },
+      editBtnClick: () => {
+        if (element) {
```
Comment (tombenezra) [inline 7-React/src/components/TodoItem/TodoItem.driver.tsx]: if the element doesn't exist, nothing will happen. this could be problematic if someone else will use this driver and expect the click to happen but no indication is given

**tl;dr** throw an error?

File: 7-React/src/components/TodoItem/TodoItem.driver.tsx [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    };
+  };
+  private deleteBtn = () => {
+    const element = this.wrapper?.queryByTestId(dataHooks.deleteBtn);
+    return {
+      deleteBtnExists: () => {
+        return element ? true : false;
+      },
+      deleteTaskClick: () => {
+        if (element) {
```
Comment (tombenezra) [inline 7-React/src/components/TodoItem/TodoItem.driver.tsx]: same here

File: 7-React/src/components/TodoItem/TodoItem.driver.tsx [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+        }
+      },
+    };
+  };
+  private deleteBtn = () => {
+    const element = this.wrapper?.queryByTestId(dataHooks.deleteBtn);
+    return {
+      deleteBtnExists: () => {
+        return element ? true : false;
+      },
```
Comment (tombenezra) [inline 7-React/src/components/TodoItem/TodoItem.driver.tsx]: this logic is repeated, consider exporting it to a function

File: 7-React/src/components/TodoItem/TodoItem.driver.tsx [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+          fireEvent.click(element);
+        }
+      },
+    };
+  };
+  private taskTitle = () => {
+    const element = this.wrapper?.queryByTestId(dataHooks.title);
+    return {
+      getTitle: () => {
+        return element?.textContent;
```
Comment (tombenezra) [inline 7-React/src/components/TodoItem/TodoItem.driver.tsx:38]: same here, you're kinda swallowing the error

File: 7-React/src/components/TodoItem/TodoItem.driver.tsx [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+    return {
+      checkBoxBtnExists: () => {
+        return element ? true : false;
+      },
+      markAsDone: () => {
+        clickCheckBox()
+      },
+      unMarkAsDone: () => {
+        clickCheckBox();
+      },
```
Comment (tombenezra) [inline 7-React/src/components/TodoItem/TodoItem.driver.tsx]: you're assuming someone will call `markAsDone` first? either check the status before clicking (so that `markAsDone` will only succeed if it actually marked an item as done).
another option is to unify them into `toggleDone` or similar (your call)

File: 7-React/src/components/TodoItem/TodoItem.driver.tsx [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+      hasCrossLine: () => {
+        return element?.classList.contains('done-task');
+      },
+    };
+  };
+
+  given = {
+    name: (name: string) => {
```
Comment (tombenezra) [inline 7-React/src/components/TodoItem/TodoItem.driver.tsx]: i suggest being consistent with naming, you're referring this input as description/name/title - stick to one

File: 7-React/src/components/TodoItem/TodoItem.spec.tsx [hunk 15]
```diff
@@ -0,0 +1,48 @@
+import TodoItemDriver from "./TodoItem.driver";
```
Comment (tombenezra) [inline 7-React/src/components/TodoItem/TodoItem.spec.tsx:1]: tests are well written and structured 👏

**Review by tombenezra — CHANGES_REQUESTED** [review]
